src/neuron/container/soa_container.hpp

```diff
@@ -1233,7 +1357,18 @@ struct soa {
     }
 
     /**
-     * @brief Flag for get_sorted_token(), mark_as_unsorted() and is_sorted().
+     * @brief Mutex to protect m_frozen_count and m_sorted.
+     *
+     * Ideally the methods that touch these would only be called before
+     * multiple worker threads have been spun up, but in practice methods such
+     * as nrn_ensure_model_data_are_sorted() are called from within
+     * multithreaded regions because neuron::model_sorted_token const& is not
+     * explicitly passed everywhere.
+     */
+    mutable std::mutex m_mut{};
```

Comment:
To me it's not fully clear what we're locking. Is it just the values `m_frozen` and `m_sorted`; or are we locking entire operations that can change the state of those variables. If we're locking "transactions" that change the value of those operations, why is locking at level of these methods sufficient?

For example `mark_as_sorted()` would seem like an operation that would be preceded by some operation that causes the `soa` to be sorted. However, since one would be releasing the lock between the two operations, we wouldn't be locking the complete transaction.

If we're only locking changes to these two values, we could put then into a struct, together with their mutex to make it more explicit what the mutex protects.

Reply:
Firstly, I agree that not all combinations of operations on `soa` containers are currently fully thread-safe.

Regarding `mark_as_sorted()` and friends, this does have a signature that takes a token as an argument, which is basically the mechanism by which we can have transactions that involve multiple operations:
```c++
auto token = soa_data.issue_frozen_token();
// calculate permutation vector, which relies on indices in soa_data being fixed
soa_data.apply_reverse_permutation(perm, token);
```
even though the mutex is not locked between the two calls, `m_frozen_count` (which is protected by the mutex) is nonzero, which prevents any operations on the container that would invalidate indices (which is the relevant thing in this example).

Reply:
Why can't we remove `mark_as_sorted()` (the variant without a token)? It feels like something that'll trick people into writing buggy code, e.g. unless you know that the execution is single threaded I don't see how `mark_as_sorted()` is useful.

Reply:
`mark_as_sorted()` without arguments is now gone.